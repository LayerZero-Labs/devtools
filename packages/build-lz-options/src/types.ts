import { OptionType } from '@layerzerolabs/lz-v2-utilities'

/**
 * Used to render OptionType input from the user.
 */
export interface OptionTypeOption {
    id: OptionType
    label: string
}

/**
 * Input OptionType selection.
 */
export interface OptionTypeInput {
    type: OptionTypeOption
}

/**
 * The result of building an Option.
 */
export interface OptionOutput {
    hex: string
    warning: string[]
}

/**
 * Summary of OptionType.TYPE_1.
 */
export interface OptionType1Summary {
    gasLimit: string
}

/**
 * Summary of OptionType.TYPE_2.
 */
export interface OptionType2Summary extends OptionType1Summary {
    nativeDropAmount: string
    nativeDropAddress: string
}
