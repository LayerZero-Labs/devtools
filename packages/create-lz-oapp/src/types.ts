export interface Config {
    destination: string
    example: Example
    packageManager: PackageManager
}

export interface Example {
    id: string
    label: string
    repository: string
    directory?: string
    ref?: string
}

export interface PackageManager {
    id: string
    executable: string
    args: string[]
    label: string
}
