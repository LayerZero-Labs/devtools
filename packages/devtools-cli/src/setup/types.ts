import type { OmniSDKFactory, OmniSignerFactory } from '@layerzerolabs/devtools'

export interface CLISetup {
    createSdk: OmniSDKFactory
    createSigner: OmniSignerFactory
}
