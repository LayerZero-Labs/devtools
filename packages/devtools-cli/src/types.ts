import type {
    Configurator,
    Factory,
    IOmniSDK,
    OmniGraph,
    OmniSDKFactory,
    OmniSignerFactory,
} from '@layerzerolabs/devtools'

export interface CLISetup<TOmniGraph extends OmniGraph = OmniGraph<unknown, unknown>, TOmniSDK = IOmniSDK> {
    createSdk: OmniSDKFactory<TOmniSDK>
    createSigner: OmniSignerFactory
    configure?: Configurator<TOmniGraph, TOmniSDK>
    loadConfig?: Factory<[configPath: string], TOmniGraph>
}
