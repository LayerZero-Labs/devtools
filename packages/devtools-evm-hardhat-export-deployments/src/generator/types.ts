import { Either } from 'fp-ts/lib/Either'

export interface OutputFile {
    path: string
    content: string
}

export type CodeGenerator = (deploymentFilePaths: string[]) => Either<Error, OutputFile[]>
