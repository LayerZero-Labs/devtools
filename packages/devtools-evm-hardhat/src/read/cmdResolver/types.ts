import type { BlockNumberTimeMarker, ResolvedTimestampTimeMarker, TimestampTimeMarker } from '@/read/types'

export interface ICommandResolverSdk {
    extractTimeMarkers(command: string): Promise<{
        blockNumberTimeMarkers: BlockNumberTimeMarker[]
        timestampTimeMarkers: TimestampTimeMarker[]
    }>

    resolveCmd(command: string, timeMarkers: ResolvedTimestampTimeMarker[]): Promise<string>
}

export * from './chain/types'
