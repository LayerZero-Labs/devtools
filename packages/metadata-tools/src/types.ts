import type { OmniPointHardhat } from '@layerzerolabs/devtools-evm-hardhat'
import type { OAppEnforcedOption } from '@layerzerolabs/ua-devtools'

// [srcContract, dstContract, [requiredDVNs, [optionalDVNs, threshold]], [srcToDstConfirmations, dstToSrcConfirmations]], [enforcedOptionsSrcToDst, enforcedOptionsDstToSrc]
export type TwoWayConfig = [
    OmniPointHardhat, // srcContract
    OmniPointHardhat, // dstContract
    [string[], [string[], number] | []], // [requiredDVNs, [optionalDVNs, threshold]]
    [number, number | undefined], // [srcToDstConfirmations, dstToSrcConfirmations]
    [OAppEnforcedOption[] | undefined, OAppEnforcedOption[] | undefined], // [enforcedOptionsSrcToDst, enforcedOptionsDstToSrc]
]

export interface IMetadata {
    [key: string]: {
        created: string
        updated: string
        tableName: string
        environment: string
        blockExplorers?: { url: string }[]
        deployments?: {
            eid: string
            chainKey: string
            stage: string
            version: number
            endpoint?: { address: string }
            relayerV2?: { address: string }
            ultraLightNodeV2?: { address: string }
            nonceContract?: { address: string }
            executor?: { address: string; pda?: string }
            deadDVN?: { address: string }
            endpointV2?: { address: string }
            sendUln302?: { address: string }
            lzExecutor?: { address: string }
            sendUln301?: { address: string }
            receiveUln301?: { address: string }
            receiveUln302?: { address: string }
        }[]
        chainDetails?: {
            chainType: string
            chainKey: string
            nativeChainId?: number
            chainLayer: string
            chainStack?: string
            nativeCurrency: {
                name?: string
                symbol: string
                cgId?: string
                cmcId: number
                decimals: number
            }
            cgNetworkId?: string
            shortName?: string
            mainnetChainName?: string
            name?: string
        }
        dvns?: {
            [address: string]: {
                version: number
                canonicalName: string
                id: string
                deprecated?: boolean
                lzReadCompatible?: boolean
            }
        }
        rpcs?: { url: string; weight?: number }[]
        addressToOApp?: {
            [address: string]: {
                id: string
                canonicalName: string
                type?: string
            }
        }
        chainName: string
        tokens?: {
            [address: string]: {
                symbol: string
                cgId?: string
                cmcId?: number
                type: string
                decimals: number
                peggedTo?: {
                    symbol: string
                    chainName: string
                    address: string
                    programaticallyPegged?: boolean
                }
            }
        }
        chainKey: string
    }
}
