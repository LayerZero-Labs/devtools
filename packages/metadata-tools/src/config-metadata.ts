import type { OmniEdgeHardhat } from '@layerzerolabs/devtools-evm-hardhat'
import type { OAppEdgeConfig } from '@layerzerolabs/ua-devtools'
import { IMetadata } from './types'
import { TwoWayConfig } from './types'

const METADATA_URL = process.env.LZ_METADATA_URL || 'https://metadata.layerzero-api.com/v1/metadata'

function getEndpointIdDeployment(eid: number, metadata: IMetadata) {
    const srcEidString = eid.toString()
    for (const objectKey in metadata) {
        const entry = metadata[objectKey]

        if (typeof entry?.deployments !== 'undefined') {
            for (const deployment of entry.deployments) {
                if (srcEidString === deployment.eid) {
                    return deployment
                }
            }
        }
    }

    throw new Error(`Can't find endpoint with eid: "${eid}",`)
}

export function DVNsToAddresses(dvns: string[], chainKey: string, metadata: IMetadata) {
    if (dvns.length === 0) {
        return []
    }

    const dvnAddresses: string[] = []
    const seenDVNs = new Set<string>()

    if (!metadata[chainKey]?.dvns) {
        throw new Error(`Can't find DVNs for chainKey: "${chainKey}".`)
    }

    const metadataDVNs = Object.entries(metadata[chainKey].dvns)

    for (const dvn of dvns) {
        if (seenDVNs.has(dvn)) {
            throw new Error(`Duplicate DVN name found: "${dvn}".`)
        }
        seenDVNs.add(dvn)

        let i = 0
        for (const [dvnAddress, dvnDetails] of metadataDVNs) {
            if (
                !dvnDetails.deprecated &&
                dvnDetails.canonicalName === dvn &&
                !dvnDetails.lzReadCompatible &&
                dvnDetails.version === 2
            ) {
                dvnAddresses.push(dvnAddress)
                break
            }

            if (i === metadataDVNs.length - 1) {
                throw new Error(
                    `Can't find DVN: "${dvn}" on chainKey: "${chainKey}". Double check you're using valid DVN canonical name (not an address).`
                )
            }

            i++
        }
    }

    if (dvns.length !== dvnAddresses.length) {
        throw new Error(`Can't find all DVNs: "${dvns.join(', ')}".`)
    }

    return dvnAddresses.sort()
}

function isSolanaDeployment(deployment: { chainKey: string; executor?: { pda?: string; address?: string } }) {
    return deployment.chainKey.startsWith('solana')
}

export async function translatePathwayToConfig(
    pathway: TwoWayConfig,
    metadata: IMetadata
): Promise<OmniEdgeHardhat<OAppEdgeConfig | undefined>[]> {
    const configs: OmniEdgeHardhat<OAppEdgeConfig | undefined>[] = []

    const sourceContract = pathway[0]
    const destinationContract = pathway[1]
    const [requiredDVNs, optionalDVNConfig] = pathway[2]
    const [sourceToDestinationConfirmations, destinationToSourceConfirmations] = pathway[3]
    const [enforcedOptionsSrcToDst, enforcedOptionsDstToSrc] = pathway[4]

    const optionalDVNs = optionalDVNConfig[0]
    const optionalDVNThreshold = optionalDVNConfig[1] || 0

    if (optionalDVNThreshold > (optionalDVNs?.length || 0)) {
        throw new Error(`Optional DVN threshold is greater than the number of optional DVNs.`)
    }

    const sourceLZDeployment = getEndpointIdDeployment(sourceContract.eid, metadata)
    const destinationLZDeployment = getEndpointIdDeployment(destinationContract.eid, metadata)

    const sourceExecutor = isSolanaDeployment(sourceLZDeployment)
        ? sourceLZDeployment.executor?.pda
        : sourceLZDeployment.executor?.address

    if (!sourceExecutor) {
        throw new Error(`Can't find executor for source endpoint with eid: "${sourceContract.eid}".`)
    }

    const sourceRequiredDVNs = DVNsToAddresses(requiredDVNs, sourceLZDeployment.chainKey, metadata)
    const destinationRequiredDVNs = DVNsToAddresses(requiredDVNs, destinationLZDeployment.chainKey, metadata)

    let sourceOptionalDVNs: string[] = []
    let destinationOptionalDVNs: string[] = []

    if (optionalDVNs) {
        sourceOptionalDVNs = DVNsToAddresses(optionalDVNs, sourceLZDeployment.chainKey, metadata)
        destinationOptionalDVNs = DVNsToAddresses(optionalDVNs, destinationLZDeployment.chainKey, metadata)
    }

    if (!sourceLZDeployment.sendUln302 || !sourceLZDeployment.receiveUln302 || !sourceLZDeployment.executor) {
        throw new Error(
            `Can't find sendUln302, receiveUln302 or executor for source endpoint with eid: "${sourceContract.eid}".`
        )
    }

    if (
        !destinationLZDeployment.sendUln302 ||
        !destinationLZDeployment.receiveUln302 ||
        !destinationLZDeployment.executor
    ) {
        throw new Error(
            `Can't find sendUln302, receiveUln302 or executor for destination endpoint with eid: "${destinationContract.eid}".`
        )
    }

    const sourceToDestinationConfig: OmniEdgeHardhat<OAppEdgeConfig> = {
        from: sourceContract,
        to: destinationContract,
        config: {
            sendLibrary: sourceLZDeployment.sendUln302.address,
            receiveLibraryConfig: {
                receiveLibrary: sourceLZDeployment.receiveUln302.address,
                gracePeriod: BigInt(0),
            },
            sendConfig: {
                executorConfig: {
                    maxMessageSize: 10000,
                    executor: sourceExecutor,
                },
                ulnConfig: {
                    confirmations: BigInt(sourceToDestinationConfirmations),
                    requiredDVNs: sourceRequiredDVNs,
                    optionalDVNs: sourceOptionalDVNs,
                    optionalDVNThreshold,
                },
            },
            enforcedOptions: enforcedOptionsSrcToDst,
        },
    }

    const destinationToSourceConfig: OmniEdgeHardhat<OAppEdgeConfig> = {
        from: destinationContract,
        to: sourceContract,
        config: {
            sendLibrary: destinationLZDeployment.sendUln302.address,
            receiveLibraryConfig: {
                receiveLibrary: destinationLZDeployment.receiveUln302.address,
                gracePeriod: BigInt(0),
            },
            receiveConfig: {
                ulnConfig: {
                    confirmations: BigInt(sourceToDestinationConfirmations),
                    requiredDVNs: destinationRequiredDVNs,
                    optionalDVNs: destinationOptionalDVNs,
                    optionalDVNThreshold,
                },
            },
        },
    }

    if (destinationToSourceConfirmations) {
        const destinationExecutor = isSolanaDeployment(destinationLZDeployment)
            ? destinationLZDeployment.executor?.pda
            : destinationLZDeployment.executor?.address

        if (!destinationExecutor) {
            throw new Error(`Can't find executor for destination endpoint with eid: "${destinationContract.eid}".`)
        }

        sourceToDestinationConfig.config.receiveConfig = {
            ulnConfig: {
                confirmations: BigInt(destinationToSourceConfirmations),
                requiredDVNs: sourceRequiredDVNs,
                optionalDVNs: sourceOptionalDVNs,
                optionalDVNThreshold,
            },
        }

        destinationToSourceConfig.config.enforcedOptions = enforcedOptionsDstToSrc

        destinationToSourceConfig.config.sendConfig = {
            executorConfig: {
                maxMessageSize: 10000,
                executor: destinationExecutor,
            },
            ulnConfig: {
                confirmations: BigInt(destinationToSourceConfirmations),
                requiredDVNs: destinationRequiredDVNs,
                optionalDVNs: destinationOptionalDVNs,
                optionalDVNThreshold,
            },
        }
    }

    configs.push(sourceToDestinationConfig)
    configs.push(destinationToSourceConfig)

    return configs
}

export async function generateConnectionsConfig(pathways: TwoWayConfig[]) {
    const metadata = (await fetch(METADATA_URL).then((res) => res.json())) as IMetadata
    const connections: OmniEdgeHardhat<OAppEdgeConfig | undefined>[] = []

    for (const pathway of pathways) {
        connections.push(...(await translatePathwayToConfig(pathway, metadata)))
    }

    return connections
}
