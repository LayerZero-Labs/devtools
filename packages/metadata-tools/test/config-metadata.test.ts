import { DVNsToAddresses, translatePathwayToConfig } from '@/config-metadata'
import { IMetadata } from '@/types'

import fujiMetadata from './data/fuji.json'
import polygonMainnetMetadata from './data/polygon-mainnet.json'
import solanaMainnetMetadata from './data/solana-mainnet.json'
import solanaTestnetMetadata from './data/solana-testnet.json'

describe('config-metadata', () => {
    const metadata: IMetadata = {
        fuji: fujiMetadata,
        solana: solanaMainnetMetadata,
        polygon: polygonMainnetMetadata,
        'solana-testnet': solanaTestnetMetadata,
    }

    describe('translatePathwayToConfig', () => {
        it('should be able to translate a pathway to a config', async () => {
            const avalancheContract = {
                eid: 40106,
                contractName: 'MyOFT',
            }

            const solanaContract = {
                eid: 40168,
                address: 'HBTWw2VKNLuDBjg9e5dArxo5axJRX8csCEBcCo3CFdAy',
            }

            const translatedConfig = await translatePathwayToConfig(
                [avalancheContract, solanaContract, [['LayerZero Labs'], []], [1, 1], [undefined, undefined]],
                metadata
            )

            expect(translatedConfig).toMatchSnapshot()
        })
    })

    describe('DVNsToAddresses', () => {
        it('returns an empty array if no DVNs are provided', () => {
            expect(DVNsToAddresses([], 'fuji', metadata)).toStrictEqual([])
        })

        it('should correctly parse DVN addresses', () => {
            expect(DVNsToAddresses(['LayerZero Labs'], 'fuji', metadata)).toStrictEqual([
                '0x9f0e79aeb198750f963b6f30b99d87c6ee5a0467',
            ])
        })

        it('sorts the EVM addresses', () => {
            expect(
                DVNsToAddresses(['LayerZero Labs', 'Nethermind', 'Google', 'Gitcoin'], 'fuji', metadata)
            ).toStrictEqual([
                '0x071fbf35b35d48afc3edf84f0397980c25531560',
                '0x7883f83ea40a56137a63baf93bfee5b9b8c1c447',
                '0x9f0e79aeb198750f963b6f30b99d87c6ee5a0467',
                '0xa4652582077afc447ea7c9e984d656ee4963fe95',
            ])
        })

        it('sorts the Solana addresses', () => {
            expect(
                DVNsToAddresses(['LayerZero Labs', 'Nethermind', 'Paxos', 'Horizen', 'Google'], 'solana', metadata)
            ).toStrictEqual([
                '4HxXbLv37XrivKukEbofybpHr7C8HUGJzd4B5T9USpGh',
                '4VDjp6XQaxoZf5RGwiPU9NR1EXSZn2TP4ATMmiSzLfhb',
                'F7gu9kLcpn4bSTZn183mhn2RXUuMy7zckdxJZdUjuALw',
                'GPjyWr8vCotGuFubDpTxDxy9Vj1ZeEN4F2dwRmFiaGab',
                'HR9NQKK1ynW9NzgdM37dU5CBtqRHTukmbMKS7qkwSkHX',
            ])
        })

        it('should not support hex string addresses', () => {
            expect(() =>
                DVNsToAddresses(['0xa4652582077afc447ea7c9e984d656ee4963fe95', 'LayerZero Labs'], 'fuji', metadata)
            ).toThrow(
                `Can't find DVN: "0xa4652582077afc447ea7c9e984d656ee4963fe95" on chainKey: "fuji". Double check you're using valid DVN canonical name (not an address).`
            )
        })

        it('should not support Solana addresses', () => {
            expect(() => DVNsToAddresses(['4VDjp6XQaxoZf5RGwiPU9NR1EXSZn2TP4ATMmiSzLfhb'], 'solana', metadata)).toThrow(
                `Can't find DVN: "4VDjp6XQaxoZf5RGwiPU9NR1EXSZn2TP4ATMmiSzLfhb" on chainKey: "solana". Double check you're using valid DVN canonical name (not an address).`
            )
        })

        it('should throw error for invalid chain key', () => {
            expect(() => DVNsToAddresses(['LayerZero Labs'], 'invalid_chain', metadata)).toThrow(
                `Can't find DVNs for chainKey: "invalid_chain".`
            )
        })

        it('should throw error for non-existent DVN name', () => {
            expect(() => DVNsToAddresses(['Non Existent DVN'], 'fuji', metadata)).toThrow(
                `Can't find DVN: "Non Existent DVN" on chainKey: "fuji". Double check you're using valid DVN canonical name (not an address).`
            )
        })

        it('should be case sensitive for DVN names', () => {
            expect(() => DVNsToAddresses(['layerzero labs'], 'fuji', metadata)).toThrow(
                `Can't find DVN: "layerzero labs" on chainKey: "fuji". Double check you're using valid DVN canonical name (not an address).`
            )
        })

        it('should handle duplicate DVN names', () => {
            expect(() => DVNsToAddresses(['LayerZero Labs', 'LayerZero Labs'], 'fuji', metadata)).toThrow(
                `Duplicate DVN name found: "LayerZero Labs".`
            )
        })

        it('correctly takes into account the version of the DVN', () => {
            expect(DVNsToAddresses(['Polyhedra'], 'polygon', metadata)).toStrictEqual([
                '0x8ddf05f9a5c488b4973897e278b58895bf87cb24',
            ])
        })

        it('should ignore deprecated DVNs incase of multiple DVNs', () => {
            expect(DVNsToAddresses(['Bitgo'], 'fuji', metadata)).toStrictEqual([
                '0xa1d84e5576299acda9ffed53195eadbe60d48e83',
            ])
        })

        it('should throw error if all DVNs are deprecated', () => {
            expect(() => DVNsToAddresses(['Deprec'], 'fuji', metadata)).toThrow(
                `Can't find DVN: "Deprec" on chainKey: "fuji". Double check you're using valid DVN canonical name (not an address).`
            )
        })
    })
})
