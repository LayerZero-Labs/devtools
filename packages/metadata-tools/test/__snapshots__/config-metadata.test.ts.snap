// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config-metadata translatePathwayToConfig should be able to translate a pathway to a config 1`] = `
[
  {
    "config": {
      "enforcedOptions": undefined,
      "receiveConfig": {
        "ulnConfig": {
          "confirmations": 1n,
          "optionalDVNThreshold": 0,
          "optionalDVNs": [],
          "requiredDVNs": [
            "0x9f0e79aeb198750f963b6f30b99d87c6ee5a0467",
          ],
        },
      },
      "receiveLibraryConfig": {
        "gracePeriod": 0n,
        "receiveLibrary": "0x819F0FAF2cb1Fba15b9cB24c9A2BDaDb0f895daf",
      },
      "sendConfig": {
        "executorConfig": {
          "executor": "0xa7BFA9D51032F82D649A501B6a1f922FC2f7d4e3",
          "maxMessageSize": 10000,
        },
        "ulnConfig": {
          "confirmations": 1n,
          "optionalDVNThreshold": 0,
          "optionalDVNs": [],
          "requiredDVNs": [
            "0x9f0e79aeb198750f963b6f30b99d87c6ee5a0467",
          ],
        },
      },
      "sendLibrary": "0x69BF5f48d2072DfeBc670A1D19dff91D0F4E8170",
    },
    "from": {
      "contractName": "MyOFT",
      "eid": 40106,
    },
    "to": {
      "address": "HBTWw2VKNLuDBjg9e5dArxo5axJRX8csCEBcCo3CFdAy",
      "eid": 40168,
    },
  },
  {
    "config": {
      "enforcedOptions": undefined,
      "receiveConfig": {
        "ulnConfig": {
          "confirmations": 1n,
          "optionalDVNThreshold": 0,
          "optionalDVNs": [],
          "requiredDVNs": [
            "4VDjp6XQaxoZf5RGwiPU9NR1EXSZn2TP4ATMmiSzLfhb",
          ],
        },
      },
      "receiveLibraryConfig": {
        "gracePeriod": 0n,
        "receiveLibrary": "7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH",
      },
      "sendConfig": {
        "executorConfig": {
          "executor": "AwrbHeCyniXaQhiJZkLhgWdUCteeWSGaSN1sTfLiY7xK",
          "maxMessageSize": 10000,
        },
        "ulnConfig": {
          "confirmations": 1n,
          "optionalDVNThreshold": 0,
          "optionalDVNs": [],
          "requiredDVNs": [
            "4VDjp6XQaxoZf5RGwiPU9NR1EXSZn2TP4ATMmiSzLfhb",
          ],
        },
      },
      "sendLibrary": "7a4WjyR8VZ7yZz5XJAKm39BUGn5iT9CKcv2pmG9tdXVH",
    },
    "from": {
      "address": "HBTWw2VKNLuDBjg9e5dArxo5axJRX8csCEBcCo3CFdAy",
      "eid": 40168,
    },
    "to": {
      "contractName": "MyOFT",
      "eid": 40106,
    },
  },
]
`;
